models/modules/loss.py

```diff
@@ -394,6 +398,362 @@ def compute_loss_G(self, netD, real, fake):
+                        "fake_bottleneck",
+                        "cutmix",
+                    ],
+                ),
+            )
+        else:
+            self.viz.line(
+                Y=[
+                    [
+                        loss_D_pixel,
+                        loss_D_bottleneck,
+                        loss_pred_real_pixel,
+                        loss_pred_real_bottleneck,
+                        loss_D_fake_pixel,
+                        loss_D_fake_bottleneck,
+                        loss_cutmix_pixel,
+                    ]
+                ],
+                X=[self.iters],
+                win=self.viz_window_cutmix,
+                update="append",
+            )
+
+        self.iters += 1
+
+    def plot_cutmix(self, real, fake, cutmix_img):
+        for i in range(real.size(0)):
+            windom_id = "cutmix_figure_batch_{}".format(i)
+            self.viz.close(win=windom_id)
+            combined = torch.cat((real[i], fake[i], cutmix_img[i]), 2)
+            combined = (combined - combined.min()) / (combined.max() - combined.min())
+
+            combined = combined.cpu().detach().numpy().transpose(1, 2, 0)
+            combined = (combined * 255).astype(np.uint8)
+            self.viz.image(
+                combined.transpose(2, 0, 1),
+                opts=dict(
+                    title="Batch {}".format(i + 1), caption="Real | Fake | Cutmix"
+                ),
+                win=windom_id,
+            )
+            time.sleep(5)
+
+    def compute_loss_D(self, netD, real, fake, fake_2):
+        """Calculate GAN loss for the discriminator
+        Parameters:
+            netD (network)      -- the discriminator D
+            real (tensor array) -- real images
+            fake (tensor array) -- images generated by a generator
+        Return the discriminator loss.
+        We also call loss_D.backward() to calculate the gradients.
+        """
+        super().compute_loss_D(netD, real, fake, fake_2)
+
+        # Real
+        pred_real_pixel, pred_real_bottleneck = netD(self.real)
+
+        loss_pred_real_pixel = self.criterionGAN(pred_real_pixel, True)
+        loss_pred_real_bottleneck = self.criterionGAN(pred_real_bottleneck, True)
+        self.loss_D_real = loss_pred_real_pixel + loss_pred_real_bottleneck
+
+        # Fake
+
+        lambda_loss = 0.5
+
+        def cutmix_real_fake_pairwise(real, fake):
+            assert (
+                self.real.size() == self.fake.size()
+            ), "Real and fake images should have the same dimensions."
+            masks = torch.stack(
+                [
+                    CutMix(self.real.size(2)).to(self.real.device)
+                    for _ in range(self.real.size(0))
+                ]
+            )
+            masks = masks.unsqueeze(1)
+            mixed_images = masks * fake + (1 - masks) * real
+            return mixed_images, masks
+
+        train_use_cutmix = True  # change this into True to use cutmix
```

Comment:
We need to make this an option, in `options/train_options.py` and then pass it as argument to `DualDiscriminatorGANLoss(DiscriminatorLoss)` constructor.

Reply:
ok, I'll make this an option.

Reply:
it is done